examples/htool.c

```diff
@@ -251,6 +251,413 @@ static int do_target_reset(uint32_t reset_option) {
+  }
+
+  int result = 0;
+  struct {
+    const char* source_file;
+  } args;
+  if (htool_get_param_string(inv, "source-file", &args.source_file)) {
+    return -1;
+  }
+
+  printf("1. Sending Initiate\n");
+  if (!dry_run) {
+    result = send_payload_update_command(dev, 0, 0, 0);
+    if (result != 0) {
+      printf("sending INITIATE failed.\n");
+      return -1;
+    }
+  }
+
+  printf("source_file: %s\n", args.source_file);
+
+  int fd = open(args.source_file, O_RDONLY, 0);
+  if (fd == -1) {
+    fprintf(stderr, "Error opening file %s: %s\n", args.source_file,
+            strerror(errno));
+    return -1;
+  }
+
+  const size_t buf_size = sizeof(struct payload_update_packet) + 2048;
+  uint8_t buffer[buf_size];
+
+  struct stat statbuf;
+  if (fstat(fd, &statbuf)) {
+    fprintf(stderr, "fstat error: %s\n", strerror(errno));
+    goto cleanup1;
+  }
+  if (statbuf.st_size > SIZE_MAX) {
+    fprintf(stderr, "file to large\n");
+    goto cleanup1;
+  }
+
+  size_t file_size = statbuf.st_size;
+  printf("File size: %zd\n", file_size);
+  uint8_t* file_data = mmap(NULL, file_size, PROT_READ, MAP_PRIVATE, fd, 0);
+
+  printf("2. Sending Data\n");
+  const size_t max_packet_size = 1024 - sizeof(struct payload_update_packet) -
+                                 sizeof(struct ec_host_request);
+
+  struct payload_update_packet* request1 = &buffer;
+  uint8_t* request1_payload = &buffer[sizeof(struct payload_update_packet)];
+  size_t rlen = 0;
+  uint8_t response;
+  int idx = 0;
+  int offset1 = 0;
+  int size = file_size;
+  uint8_t* data = file_data;
+  while (size > 0) {
+    size_t skip = size_to_skip(data, size);
+    data += skip;
+    size -= skip;
+    offset1 += skip;
+    if (size == 0) {
+      break;
+    }
+    size_t max_size_to_send = (size < max_packet_size ? size : max_packet_size);
+    size_t to_send = size_to_send(data, max_size_to_send);
+    if (to_send) {
+      request1->offset = offset1;
+      request1->len = to_send;
+      request1->type = 1;  // PAYLOAD_UPDATE_CONTINUE
+      memcpy(request1_payload, data, to_send);
+      result = 0;
+    RETRY:
+      if (!dry_run) {
+        result = htool_exec_hostcmd(
+            dev, EC_CMD_BOARD_SPECIFIC_BASE + EC_PRV_CMD_HOTH_PAYLOAD_UPDATE,
+            NULL, request1, sizeof(*request1) + request1->len, &response,
+            sizeof(response), &rlen);
+      }
+      if (result != 0) {
+        printf("Error, result=%d. will retry\n", result);
+        goto RETRY;
+      }
+      printf("[Frame %d] offset=%zd len=%zd, response=%u ", idx, offset1,
+             to_send, response);
+      for (int i = 0; i < 40; i++) {
+        printf("%02X ", buffer[i]);
+      }
+      printf("\n");
+      data += max_size_to_send;
+      size -= max_size_to_send;
+      offset1 += max_size_to_send;
+    }
+
+    idx++;
+  }
+
+  result = send_payload_update_command(dev, 2, 0, 0);  // Finalize
+  if (result != 0) {
+    printf("sending FINALIZE failed.\n");
+    return -1;
```

Comment:
Return here skips the `cleanup2` and `cleanup` blocks.  (I'm not especially fond of the `goto cleanup_label` pattern, since it often causes this sort of bug, but the `..._internal` function pattern is noisy.)